privacy-protections-popup/privacy-protections-popup-impl/src/main/java/com/duckduckgo/privacyprotectionspopup/impl/PrivacyProtectionsPopupManagerImpl.kt

```diff
@@ -0,0 +1,267 @@
+import com.duckduckgo.privacyprotectionspopup.impl.db.PopupDismissDomainRepository
+import com.duckduckgo.privacyprotectionspopup.impl.db.ToggleUsageTimestampRepository
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import org.threeten.bp.Duration
+import org.threeten.bp.Instant
+
+@ContributesBinding(FragmentScope::class)
+class PrivacyProtectionsPopupManagerImpl @Inject constructor(
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+    private val featureAvailability: PrivacyProtectionsPopupFeatureAvailability,
+    private val protectionsStateProvider: ProtectionsStateProvider,
+    private val timeProvider: TimeProvider,
+    private val popupDismissDomainRepository: PopupDismissDomainRepository,
+    private val userAllowListRepository: UserAllowListRepository,
+    private val toggleUsageTimestampRepository: ToggleUsageTimestampRepository,
+    private val duckDuckGoUrlDetector: DuckDuckGoUrlDetector,
+) : PrivacyProtectionsPopupManager {
+
+    private val state = MutableStateFlow(
+        State(
+            featureAvailable = false,
+            protectionsEnabled = null,
+            domain = null,
+            hasHttpErrorCodes = false,
+            hasBrowserError = false,
+            popupDismissed = null,
+            toggleUsed = null,
+            shouldShowPopup = false,
+        ),
+    )
+
+    private var dataLoadingJob: Job? = null
+
+    override val viewState = state
+        .map { PrivacyProtectionsPopupViewState(visible = it.shouldShowPopup) }
+        .onStart { startDataLoading() }
+        .onCompletion { stopDataLoading() }
+        .stateIn(
+            scope = appCoroutineScope,
+            started = SharingStarted.WhileSubscribed(),
+            initialValue = PrivacyProtectionsPopupViewState(visible = false),
+        )
+
+    override fun onUiEvent(event: PrivacyProtectionsPopupUiEvent) {
+        when (event) {
+            DISMISSED -> {
+                dismissPopup()
+            }
+
+            DISMISS_CLICKED -> {
+                // TODO pixel
+                dismissPopup()
+            }
+
+            DISABLE_PROTECTIONS_CLICKED -> {
+                // TODO pixel
```

Comment:
Is the pixel going to be added at some point?

Reply:
All the pixels are added in this stacked PR: https://github.com/duckduckgo/Android/pull/4114